docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,241 @@
+```luau
+type function neverending(t)
+    return neverending(t) -- note: parentheses are used here because the runtime value of types is being passed in, rather than the static annotation of types
+end
+```
+As a result, the current plan is to maintain a tightly scoped implementation, focusing on the simplest version of type functions to minimize the risk of developers crashing the analysis, and gradually enhance capabilities through successive iterations.
+
+We have considered implementing a user-configurable execution limit based on time or instruction count for reducing type functions where if a type function exceeds the limit, it fails to reduce. However, these methods are not consistently reliable for terminating type functions that take too long to execute. Time-based timeouts are dependent on CPU performance; programs that type check on fast CPUs may not type check on slower CPUs. Similarly, instruction-based timeouts are dependent on the compiler optimizations; programs that type check on one compiler version may not type check on versions without the same optimizations. We will be experimenting with various forms of limiting the execution of type functions and reserve the right to change how termination of type functions is managed.
+
+<details><summary>List of illegal constructs in type functions</summary>
+
+* `while` / `repeat` loops
+* invoking other type functions / regular functions / lambdas
+    * we will not (and probably never) allow type functions to call regular functions for the sake of sandboxing the runtime and analysis.
+* referring to locals / globals in the outer scope
+* global functions: `getfenv`, `setfenv`, `pcall`, `xpcall`, `require`
+* libraries: `coroutine`, `debug`, `string.gsub`
+
+Note: we are aware that for loops can cause infinite runtime. For the time being, we will not be handling this case. In the event that a developer accidentally creates an infinitely long type function, they will need to fix the type function and restart the analysis.
+
+</details>
+
+To give warnings, developers can use `print()` with custom warning messages. To fail reductions, developers can use `error()` with custom error messages. If nothing is returned by the type function, it will fail to reduce with the default message: "Failed to reduce \<Name\> type function with no return values".
+
+To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `typelib`. A `typelib` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, the `typelib` library will closely mimic the implementation of static types in Luau Analysis. Most importantly, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 
```

Comment:
Hate to bikeshed the name again, but could we call them simply `ty`? The clash with `type` global is unfortunate, but I think `typelib` is weirder than `ty` when no libraries we offer have `lib` in the name.

So essentially `s/typelib/ty/`.

```suggestion
To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `ty`. A `ty` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, the `ty` library will closely mimic the implementation of static types in Luau Analysis. Most importantly, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 
```

Reply:
When I proposed `typelib` it was also only intended to be the name of the _library_, not the name of the userdata (which I wanted to just be called `type`) anyway. Really, I kind of just want to use `type` and overload `__call` to be the builtin `type`, but barring that, I'm happy with `ty`. Other bikeshed possibilities: `rttype`, `rtty`, `typ`.

Reply:
Overloading `__call` on a library is not farfetched. `vector` RFC wants to do this, and no one seems to bat an eye.

If you say `type(type)` you'll get `"function"` in normal context and `"table"` in type function context, but no one should be concerned with that.